tests/src/test/scala/org/json4s/TypeHintsTest.scala

```diff
@@ -0,0 +1,30 @@
+package org.json4s
+import org.json4s.jackson.{JsonMethods, Serialization}
+import org.specs2.mutable.Specification
+
+class TypeHintTest extends Specification {
+
+  import TypeHintsTest._
+
+  "deserialization with type hints" should {
+
+    implicit val formats: Formats = DefaultFormats + MappedTypeHints(Map(classOf[Foo] -> "foo", classOf[Bar] -> "bar"))
+
+    "fail when the type hint is incompatible with the requested type" in {
+      val dump = Serialization.write(Foo(1))
+      (JsonMethods.parse(dump).extract[Bar]: Any) must throwA[MappingException]
```

Comment:
The issue I reported in https://github.com/json4s/json4s/issues/617 only seems to occur when the extracted value is then treated as an instance of `Any` (which is what happened when you pass it to `println`). I don't understand why this changes the behavior of `extract`, but making the change I did to `Extraction` get the behavior I expected originally.